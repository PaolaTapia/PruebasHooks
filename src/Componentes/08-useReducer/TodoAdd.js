import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({handleAddToDo}) => {
     const [{description}, handledInputChange, reset] = useForm({
          description: '',
     });   

const handleSubmit = (e) => {
     e.preventDefault(); 
     if(description.trim().length<=1){
          return; 
     }
     const newToDo = {
          id: new Date().getTime(),
          desc: 'Aprender '+ description,
          done: false
     }
     handleAddToDo(newToDo); 
     reset();     
}
     
  return (
    <>
      <h4>Agregar ToDo</h4>

      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          className="form-control"
          placeholder="Aprender..."
          autoComplete="off"
          onChange={handledInputChange}
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
