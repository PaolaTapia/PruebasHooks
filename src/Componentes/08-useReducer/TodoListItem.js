import React from 'react'; 

export const TodoListItem = ({toDos, handleDelete, handleToggle}) => {
     toDos
     return (
          <>
          {toDos.map((toDo, i) => (
          <li key={toDo.id} className="list-group-item">
               <p 
                    className={`${toDo.done && 'complete'}`}
                    onClick={ ()=>handleToggle (toDo.id)}
               >{i + 1}.{toDo.desc}</p>
               <button 
               onClick={()=>handleDelete(toDo.id)}
               className="btn btn-danger"
               >Borrar</button>
          </li>
          ))}
          </>
     )
}
