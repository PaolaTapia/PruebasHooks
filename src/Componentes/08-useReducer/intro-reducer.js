const initialState=[{
     id:1, 
     todo: 'comprar pan', 
     done: false
}]

const toDoReducer = (state= initialState, action) =>{

     if(action?.type==='agregar'){
          return [...state, action.payload]; 
     }

     return state; 
}

let toDos = toDoReducer(); 


const newToDo= {
     id:2, 
     todo: 'comprar leche', 
     done: false
}

const action={
     type: 'agregar', 
     payload: newToDo
}

toDos= toDoReducer (toDos, action); 

console.log(toDos); 