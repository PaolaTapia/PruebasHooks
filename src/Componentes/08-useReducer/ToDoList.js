import React from 'react'; 
import { TodoListItem } from './TodoListItem'; 

export const ToDoList = ({toDos,handleToggle,handleDelete}) => {
     return (         

          <ul className="list-group list-group-flush">         
          <TodoListItem  
               handleDelete={handleDelete} 
               handleToggle={handleToggle} 
               toDos={toDos} 
          />
         
          </ul>

     )
}
