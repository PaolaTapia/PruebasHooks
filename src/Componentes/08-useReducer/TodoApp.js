import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { TodoAdd } from "./TodoAdd";
import { ToDoList } from "./ToDoList";
import { toDoReducer } from "./toDoReducer";

export const TodoApp = () => { 
     const init= () => {
          return JSON.parse(localStorage.getItem('toDos')) || []; 

     }

     const [toDos, dispatch] = useReducer(toDoReducer,[], init);

     useEffect(() => {
      localStorage.setItem('toDos', JSON.stringify(toDos)); 
     }, [toDos])

     const handleDelete =(toDoId)=>{
     const action ={
          type: 'delete',
          payload: toDoId
     }
     dispatch(action);
     }

     const handleToggle =(toDoId) =>{
     dispatch({
          type: 'toggle',
          payload: toDoId
     });       
  }

  const handleAddToDo= (newToDo) => {  
     dispatch({
          type: 'add',
          payload: newToDo
     });
  }

 

  return (
    <div>
      <h1>TodoApp ({toDos.length})</h1>
      <hr />
     <div className="row">
     <div className="col-7">
          <ToDoList 
               toDos={toDos} 
               handleDelete={handleDelete} 
               handleToggle={handleToggle}
          />
      </div>
      <div className="col-5">
          <TodoAdd 
               handleAddToDo={handleAddToDo}
          />
      </div>
     </div>
    </div>
  );
};
