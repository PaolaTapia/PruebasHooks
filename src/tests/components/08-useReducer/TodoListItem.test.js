import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { TodoListItem } from '../../../Componentes/08-useReducer/TodoListItem';
import  { demoToDos } from '../../fixtures/demoTodos';

describe('pruebas en ToDoListItem', () => {
const handleDelete= jest.fn(); 
const handleToggle= jest.fn(); 

const wrapper = shallow(
     <TodoListItem 
          toDos={demoToDos}           
          handleDelete={handleDelete}
          handleToggle={handleToggle}
     />   );   

     test('Debe de mostrarse correctamente', () => {         
               expect(toJson(wrapper)).toMatchSnapshot();          
     }); 

     test('Debe de llamar la funcion handleDelete', () => {
          wrapper.find('button').at(0).simulate('click'); 
          expect(handleDelete).toHaveBeenCalledWith(demoToDos[0].id); 
         
     }); 

     test('Debe de llamar la funcion handleToggle', () => {
    
          wrapper.find('p').at(0).simulate('click'); 
          expect(handleToggle).toHaveBeenCalledWith(demoToDos[0].id); 
     }); 

     test('Debe de mostrar el texto correctamente', () => {
          const toDo=demoToDos[0].desc;          
          const id= demoToDos[0].id; 

          const p = wrapper.find('p').at(0); 
          expect(p.text().trim()).toBe(`${id}.${toDo}`);     
          
     }); 

     
     test('Debe de tener la clase complete si el ToDO.done=true', () => {                
          demoToDos[0].done=true;  
          
          const wrapper = shallow(
               <TodoListItem 
                    toDos={demoToDos}          
               />   );   
//console.log(wrapper.find('p').at(0).html()); 
               expect(wrapper.find('p').at(0).hasClass('complete')).toBe(true);            
     }); 

     test('Debe de tener dos toDos', () => {
          
          expect(wrapper.find('li').length).toBe(demoToDos.length);
      }); 
     

})
