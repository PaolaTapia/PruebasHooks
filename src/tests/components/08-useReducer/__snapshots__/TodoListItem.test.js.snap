// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en ToDoListItem Debe de mostrarse correctamente 1`] = `
<Fragment>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      1
      .
      Aprender React
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
  <li
    className="list-group-item"
    key="2"
  >
    <p
      className="false"
      onClick={[Function]}
    >
      2
      .
      Aprender Mongo
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</Fragment>
`;
